File: BloomFilter/BloomFilter.java
import java.util.BitSet;

public class BloomFilter {
    private static final int DEFAULT_SIZE = 2 << 24;
    private static final int[] seeds = new int[]{7, 11, 13, 31, 37, 61,};
    private BitSet bits = new BitSet(DEFAULT_SIZE);
    private SimpleHash[] func = new SimpleHash[seeds.length];

    public static void main(String[] args) {
        String value = "wangyaning6166@gmail.com";
        BloomFilter filter = new BloomFilter();
        System.out.println(filter.contains(value));
        filter.add(value);
        System.out.println(filter.contains(value));
    }

    public BloomFilter() {
        for (int i = 0; i < seeds.length; i++) {
            func[i] = new SimpleHash(DEFAULT_SIZE, seeds[i]);
        }
    }

    public void add(String value) {
        for (SimpleHash f : func) {
            bits.set(f.hash(value), true);
        }
    }

    public boolean contains(String value) {
        for (SimpleHash f : func) {
            if (!bits.get(f.hash(value))) {
                return false;
            }
        }

        return true;
    }
}

File: BloomFilter/BloomFilter.cpp
//
// Created by yaning on 6/20/18.
//

#include "BloomFilter.h"

void BloomFilter::add(const std::string &value) {
    for (auto &func : funcs) {
        bits.set(func.hash(value), true);
    }
}

bool BloomFilter::contains(const std::string &value) {
    for (auto &func : funcs) {
        if (!bits[func.hash(value)]) {
            return false;
        }
    }
    return true;
}

BloomFilter::BloomFilter() {
    int size = sizeof(seeds) / sizeof(int);
    for (int i = 0; i < size; ++i) {
        funcs.push_back(SimpleHash(seeds[i], DEFAULT_SIZE));
    }
}
